readme.md
## [![npm][npmjs-img]][npmjs-url] [![mit license][license-img]][license-url] [![build status][travis-img]][travis-url] [![coverage status][coveralls-img]][coveralls-url] [![deps status][daviddm-img]][daviddm-url]

> Get unique and sequential current week identifier for given valid `Date` string format or Date object.
Week #1 is starting on January 05, 1970.

## Install
```
npm i --save week-identifier
npm test
week-identifier --help
```


## API
> For more use-cases see the [tests](./test.js)

### [weekIdentifier](./index.js#L50)
> Get unique and sequential current week identifier for given valid `Date` string format

- `[date]` **{String}** every valid Date-ish string format
- `return` **{Number}**

**Example:**

```js
var weekIdentifier = require('week-identifier');

 // august 12, 2016
weekIdentifier();
//=> 2432

weekIdentifier('January 05, 1970 00:00:00');
//=> 1

weekIdentifier('January 12, 1970 00:00:00');
//=> 2

weekIdentifier(new Date('August 12, 2016'));
//=> 2432

weekIdentifier('08/12/2016');
//=> 2432

weekIdentifier('August 12, 2016');
//=> 2432

weekIdentifier(new Date('August 19, 2016'));
//=> 2433

```

### [weekIdentifier.dateFromWeek](./index.js#L86)
> Get monday date of the given week identifier or January 5, 1970 00:00:00 if weekIdentifier is <= 1.

- `[number]` **{String}** every valid number > 0
- `return` **{Date}**

**Example:**

```js
var weekIdentifier = require('week-identifier');

weekIdentifier.dateFromWeek(2433);
//=> August 15, 2016 00:00:00

weekIdentifier.dateFromWeek(1);
//=> January 05, 1970 00:00:00

```

## CLI
> You can just run `week-identifier --help` for more information


## License [![MIT license][license-img]][license-url]
Copyright (c) 2016 [Clément Billiot], [contributors][contrib-graf].  
Released under the [`MIT`][license-url] license.


[npmjs-url]: http://npm.im/week-identifier
[npmjs-img]: https://img.shields.io/npm/v/week-identifier.svg?style=flat&label=week-identifier

[coveralls-url]: https://coveralls.io/r/throll/week-identifier?branch=master
[coveralls-img]: https://img.shields.io/coveralls/throll/week-identifier.svg?style=flat

[license-url]: https://github.com/throll/week-identifier/blob/master/license.md
[license-img]: https://img.shields.io/badge/license-MIT-blue.svg?style=flat

[travis-url]: https://travis-ci.org/throll/week-identifier
[travis-img]: https://img.shields.io/travis/throll/week-identifier.svg?style=flat

[daviddm-url]: https://david-dm.org/throll/week-identifier
[daviddm-img]: https://img.shields.io/david/dev/throll/week-identifier.svg?style=flat

[author-github]: https://github.com/throll

[contrib-graf]: https://github.com/throll/week-identifier/graphs/contributors

***


history.md
## v0.1.0 / August 16, 2016
- Release v0.1.0 / npm@v0.1.0
- initial commits


test.js
/**
 * week-idnetifier <https://github.com/throll/week-identifier>
 *
 * Copyright (c) 2016 Clément Billiot, contributors.
 * Released under the MIT license.
 */

'use strict';

var assert = require('chai').assert;
var chai = require('chai');
chai.use(require('chai-datetime'));
var weekIdentifier = require('./index');

describe('week-identifier:', function() {
  it('should receive valid date format string', function(done) {
    assert.strictEqual(weekIdentifier('January 05, 1970 03:24:00'), 1);
    assert.strictEqual(weekIdentifier('January 11, 1970 23:59:59'), 1);
    assert.strictEqual(weekIdentifier('January 12, 1970 00:00:00'), 2);
    assert.strictEqual(weekIdentifier('January 12, 1970'), 2);
    assert.strictEqual(weekIdentifier('January 14, 1970'), 2);
    assert.strictEqual(weekIdentifier('01/14/1970'), 2);
    assert.strictEqual(weekIdentifier('January 20, 1970 00:00:01'), 3);
    assert.strictEqual(weekIdentifier('January 20, 2016 00:00:00'), 2403);
    assert.strictEqual(weekIdentifier('January 26, 2016 00:00:00'), 2404);
    assert.strictEqual(weekIdentifier('August 12, 2016'), 2432);
    assert.strictEqual(weekIdentifier('02/17/2012'), 2198);

    done();
  });
  it('should receive date object', function(done) {
    assert.strictEqual(weekIdentifier(new Date('January 05, 1970 03:24:00')), 1);
    assert.strictEqual(weekIdentifier(new Date('August 19, 2016')), 2433);
    done();
  });
  it('should get current week identifier when empty string format', function(done) {
    assert.strictEqual(typeof weekIdentifier(''), 'number');
    done();
  });
  it('should get current week identifier, if no valid Date format or Date Object', function(done) {
    assert.strictEqual(typeof weekIdentifier(/regex/g), 'number');
    assert.strictEqual(typeof weekIdentifier(true), 'number');
    assert.strictEqual(typeof weekIdentifier(false), 'number');
    assert.strictEqual(typeof weekIdentifier(222), 'number');
    assert.strictEqual(typeof weekIdentifier({
      obj: 123
    }), 'number');
    assert.strictEqual(typeof weekIdentifier([1, 2, 3]), 'number');
    assert.strictEqual(typeof weekIdentifier(), 'number');
    done();
  });

  it('should get date from week identifier', function(done) {
    assert.isNaN(weekIdentifier.dateFromWeek(''));
    assert.isNaN(weekIdentifier.dateFromWeek('abc'));
    assert.equalDate(weekIdentifier.dateFromWeek(-2897), new Date('January 05, 1970 00:00:00'));
    assert.equalDate(weekIdentifier.dateFromWeek('0'), new Date('January 05, 1970 00:00:00'));
    assert.equalDate(weekIdentifier.dateFromWeek('1'), new Date('January 05, 1970 00:00:00'));
    assert.equalDate(weekIdentifier.dateFromWeek('2'), new Date('January 12, 1970 00:00:00'));
    assert.equalDate(weekIdentifier.dateFromWeek('2403'), new Date('January 18, 2016 00:00:00'));
    assert.equalDate(weekIdentifier.dateFromWeek('2404'), new Date('January 25, 2016 00:00:00'));
    done();
  });
});


cli.js
#!/usr/bin/env node

/**
 * week-identifier <https://github.com/throll/week-identifier>
 *
 * Copyright (c) 2016-2017 Clément Billiot, contributors.
 * Released under the MIT license.
 */

'use strict';

var pkg = require('./package.json');
var weekIdentifier = require('./');
var argv = process.argv.slice(2);

function help() {
  console.log([
    '',
    '  ' + pkg.description,
    '',
    '  Example',
    '    $ week-identifier January 5, 1970',
    '    1',
    '    $ week-identifier August 12, 2016',
    '    2432',
    '    $ week-identifier 02/17/2012',
    '    2198',
    '    $ week-identifier',
    '    2433'
  ].join('\n'));
}

if (argv.indexOf('--help') !== -1) {
  return help();
}

if (argv.indexOf('--version') !== -1) {
  console.log(pkg.version);
  return;
}

console.log(weekIdentifier(argv.join('')));


index.js
/**
 * week-identifier <https://github.com/throll/week-identifier>
 *
 * Copyright (c) 2016 Clément Billiot, contributors.
 * Released under the MIT license.
 */

'use strict';

module.exports = weekIdentifier;
module.exports.weekIdentifier = weekIdentifier;
module.exports.dateFromWeek = dateFromWeek;

/**
 * Get unique and sequential week identifier of current date or given valid `Date` string format
 *
 * **Example:**
 *
 * ```js
 * var weekIdentifier = require('current-week-number');
 *
 * // august 12, 2016 (current date)
 * weekIdentifier();
 * //=> 2432
 *
 * weekIdentifier('January 05, 1970 03:00:00');
 * //=> 1
 *
 * weekIdentifier(new Date('August 12, 2016'));
 * //=> 2432
 *
 * weekIdentifier('08/12/2016');
 * //=> 2432
 *
 * weekIdentifier('August 12, 2016');
 * //=> 2432
 *
 * weekIdentifier(new Date('August 19, 2016'));
 * //=> 2433
 *
 * weekIdentifier('08/19/2016');
 * //=>  2433
 * ```
 *
 * @name weekIdentifier
 * @param  {String} `[date]` every valid Date-ish string format
 * @return {Number}
 * @api public
 */
function weekIdentifier(date) {
  var instance;

  if (typeof date === 'string' && date.length) {
    instance = new Date(date);
  } else if (date instanceof Date) {
    instance = date;
  } else {
    instance = new Date();
  }
  // Create a copy of this date object
  var target = new Date(instance.valueOf());
  // Starting date point for our sequence
  var lastDayOfWeekZero = new Date('January 4, 1970 23:59:59.999');
  // Number of week from our starting date
  var weekNumberdiff = Math.ceil((target.getTime() - lastDayOfWeekZero.getTime()) / (24 * 3600 * 1000 * 7));

  return weekNumberdiff;
}

/**
 * Get monday date of the given week identifier
 *
 *  * **Example:**
 *
 * ```js
 * var weekIdentifier = require('current-week-number');
 *
 * weekIdentifier.dateFromWeek(2403);
 * //=> January 18, 2016 00:00:00
 *
 * ```
 * @name weekIdentifier.dateFromWeek
 * @param  {Number} weekIdentifier
 * @return {Date} Monday of the given week identifier or January 5, 1970 00:00:00 if weekIdentifier is not > 0.
 */
function dateFromWeek(weekIdentifier) {
  if (isNaN(parseFloat(weekIdentifier))) {
    return NaN;
  } else {
    // Starting date point for our sequence
    var firstDayOfWeekOne, mondayOfWeek;
    mondayOfWeek = firstDayOfWeekOne = new Date('January 5, 1970 00:00:00');
    if (weekIdentifier > 0) {
      mondayOfWeek = new Date(((weekIdentifier - 1) * (24 * 3600 * 1000 * 7)) + firstDayOfWeekOne.getTime());
    }
    return mondayOfWeek;
  }
}
